src/qref/verification.py

```diff
@@ -0,0 +1,166 @@
+        routine: Routine or program to be verified.
+    """
+    if isinstance(routine, SchemaV1):
+        routine = routine.program
+    problems = _verify_routine_topology(routine)
+    return TopologyVerificationOutput(problems)
+
+
+def _verify_routine_topology(routine: RoutineV1) -> list[str]:
+    problems = []
+    adjacency_list = _get_adjacency_list_from_routine(routine, path=None)
+
+    problems += _find_cycles(adjacency_list)
+    problems += _find_disconnected_ports(routine)
+
+    for child in routine.children:
+        new_problems = _verify_routine_topology(child)
+        problems += new_problems
+    return problems
+
+
+def _get_adjacency_list_from_routine(
+    routine: RoutineV1, path: Optional[str]
+) -> dict[str, list[str]]:
+    """This function creates a flat graph representing one hierarchy level of a routine.
+
+    Nodes represent ports and edges represent connections (they're directed).
+    Additionaly, we add node for each children and edges coming from all the input ports
+    into the children, and from the children into all the output ports.
+    """
+    graph = defaultdict(list)
+    if path is None:
+        current_path = routine.name
+    else:
+        current_path = ".".join([path, routine.name])
+
+    # First, we go through all the connections and add them as adges to the graph
+    for connection in routine.connections:
+        source = ".".join([current_path, connection.source])
+        target = ".".join([current_path, connection.target])
+        graph[source].append(target)
+
+    # Then for each children we add an extra node and set of connections
+    for child in routine.children:
+        input_ports = []
+        output_ports = []
+
+        child_path = ".".join([current_path, child.name])
+        for port in child.ports:
+            if port.direction == "input":
+                input_ports.append(".".join([child_path, port.name]))
+            elif port.direction == "output":
+                output_ports.append(".".join([child_path, port.name]))
+
+        for input_port in input_ports:
+            graph[input_port].append(child_path)
+
+        graph[child_path] += output_ports
+
+    return graph
+
+
+def _find_cycles(adjacency_list: dict[str, list[str]]) -> list[str]:
+    # Note: it only returns the first detected cycle.
+    for node in list(adjacency_list.keys()):
+        problem = _dfs_iteration(adjacency_list, node)
+        if problem:
+            return problem
+    return []
+
+
+# def _dfs_iteration(adjacency_list, initial_node, node, visited):
+#     if node != initial_node:
+#         visited.append(node)
+#     for neighbour in adjacency_list[node]:
+#         if neighbour not in visited:
+#             if neighbour == initial_node:
+#                 return [f"Cycle detected for node: {node}. Cycle: {visited}."]
+#             problem = _dfs_iteration(adjacency_list, initial_node, neighbour, visited)
+#             if problem:
+#                 return problem
```

Comment:
**issue**: You know what to do with these lines.

Reply:
😅 Fixed!